app/src/main/java/com/b_lam/resplash/ui/photo/detail/PhotoDetailActivity.kt

```diff
@@ -310,22 +344,39 @@ class PhotoDetailActivity :
 
     private fun applyWallpaper(uri: Uri) {
         try {
-            startActivity(WallpaperManager.getInstance(this).getCropAndSetWallpaperIntent(uri))
+            startActivity(wallpaperManager.getCropAndSetWallpaperIntent(uri))
         } catch (e: IllegalArgumentException) {
-            var bitmap: Bitmap? = null
-            try {
-                bitmap = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
-                    ImageDecoder.decodeBitmap(ImageDecoder.createSource(contentResolver, uri))
-                } else {
-                    MediaStore.Images.Media.getBitmap(contentResolver, uri)
-                }
-                WallpaperManager.getInstance(applicationContext).setBitmap(bitmap)
-                toast("Wallpaper set successfully")
-            } catch (e: Exception) {
-                error("Error setting wallpaper", e)
+            viewModel.prepareBitmapFromUri(contentResolver, uri)
+            observerWallpaperBitmapResult()
+        }
+    }
+
+    private fun observerWallpaperBitmapResult() {
+        viewModel.wallpaperBitmap.observe(this) { result ->
+            if (result is Result.Success) {
+                setWallpaperWithBitmap(result.value)
+            } else {
                 toast("Failed to set wallpaper")
-            } finally {
-                bitmap?.recycle()
+            }
+        }
+    }
+
+    private fun setWallpaperWithBitmap(bitmap: Bitmap) {
+        lifecycleScope.launch {
+            var isWallpaperSet = false
+            withContext((Executors.newSingleThreadExecutor().asCoroutineDispatcher())) {
```

Comment:
Would `Dispatchers.IO` or `Dispatchers.Default` work here? Is there a reason you went with `Executors.newSingleThreadExecutor().asCoroutineDispatcher()`?

Reply:
Yes, actually. I wanted to execute the task in a different context, which is similar to creating a different thread to perform the bitmap's heavy tasking.